cluster-autoscaler/core/static_autoscaler.go

```diff
@@ -379,6 +379,8 @@ func (a *StaticAutoscaler) RunOnce(currentTime time.Time) errors.AutoscalerError
 		} else {
 			klog.V(4).Infof("Starting scale down")
 
+			scaleDown.SoftTaintUnneededNodes(allNodes)
```

Comment:
Can you add a metric for how long this takes (ie. metrics.UpdateDurationFromStart)? This potentially involves multiple API calls, so it's worth keeping track of the performance impact.

Reply:
Also - any specific reason this is called here? It's not obvious to me why IsDeleteInProgress() should stop us from soft tainting other nodes.

Reply:
I have added scaleDown:softTaintUnneeded metric with combined time of that loop.

It turns out that with default limit of 30 requests, it takes up to 10s.
Why: on scaling down, API server burts requests limit could be drained for pod removal request, after that API server limits to 10 qps, which translates to 5 soft taints (each take GET and PUT).
For each subsequent requests, around 200ms has to pass, so one tainting per 400ms.
I have changed default limit to 10 requests per autoscaler loop to keep up with bulk deletion. 

I plan to add another PR changing tainting retry loop from GET + PUT, to try PUT, then GET on failure. That will reduce number of requests.

Reply:
As for the place to put SoftTaintUnneededNodes:
- it should not slow down scaleUp (it is only soft tainting)
- there is flag calculateUnneededOnly marked "In dry run only utilization is updated" - I have put softTaint to run when it is false
- it does not run when scaleDownForbidden or ScaleDownInCooldown, as then we do not have full understanding of situation
- it does not run when IsDeleteInProgress to not conflict with given node removed from unneeded nodes during node deletion (in other words - to be sure that taint is not removed in edge cases)

Reply:
Sounds good to me. 

Another thing to consider in follow up PR - if it can take long, maybe we should also check elapsed time as loop condition?

Reply:
We could. Do you have any specific timeout in mind? Like 10s? Adding additional flag seems an overkill.

Reply:
Agreed we can go with const for a timeout. I'd say 10s is too much - ideally the whole loop should take no more than 10s (not true on large clusters, but one can always dream). Something like 3s sound more reasonable to me. If we're waiting 10 minutes to delete a node we can taint it one or two loops later.

Reply:
I have added time limiting as well